### ISSUE TITLE
Hibernating State (Not Necessarily Serialized)

### ISSUE THREAD

**Issue by sebmarkbage** [facebook/react#4594]
Relay and others currently abuses some internals to get some persistent identity of a component. This effectively is used to restore the state of a component after it has been temporarily unmounted. It is also common to abuse Flux stores for this use case. Basically, since this capability doesn't currently exists, you're encouraged to use Flux for everything just in case you need this capability later on.

The purpose of this issue is to discuss a public API for hibernating the state of a component and then restoring it once the component remounts.

The use cases can be broken down into three scenarios:

- __List Item Out of View__:  E.g. Infinite scrolling (such as "table views") where one row eventually needs to be reclaimed to save the memory used by the tree that is out of view.

- __Detail View__: In a master-detail view, clicking one item in the list switches the state of the detail view. When you click on the original view again.

- __Back/Forward Button__: You want to save a snapshot of the state when you navigate. The new view can then change the state and when you hit the back button to return to a previous state, you want to restore the original state of the subtree.

We would like to support this at least in a non-serialized form. You could imagine having an API that serializes this to JSON or some other data structure too but that's a potential follow up and not necessarily part of this.

One potential API:

```js
class View {
  state = { stateKeys: [{}, {}, {}] }
  render() {
    return <ChildView key={this.state.stateKeys[this.props.index]} />;
  }
}
```

Basically, an object is used as a key. Unlike the normal key semantics, the state of the subtree is kept in memory indefinitely. We use a WeakMap to keep the state. If the object ever goes away, the GC will collect the state of that subtree. This solves all three use cases.

**Comment by jimfb** [comment]
Personally, I'm still a fan of the state-hoisting pattern:

```javascript
class View {
  state = { childStates: [ new ChildViewInternalState(), new ChildViewInternalState(), new ChildViewInternalState()] }
  render() {
    return <ChildView state={this.state.childStates[this.props.index]} />;
  }
```

Component implementations provide a black-box state object which may be retained by the parent.  If the parent does not provide the component's state, the component will initialize one internally (in which case, the parent forfeits the option to preserve that child component's internal state).

This also solves all three use cases, without any special logic within the core and without increasing our API surface area.  It has added benefits (for instance, allowing the parent to query a component's internal state without getting a ref to the component, if the implementor chooses to expose that internal information as part of their public API).  Serializability is decided by the implementor of ChildView.

IMO, it is sufficient to more aggressively teach this design pattern.

**Comment by sebmarkbage** [comment]
@jimfb See #4595 for why that's not relevant to this task.

**Comment by jimfb** [comment]
Yep, already cross-referenced those issues.  Continuing the discussion there.

**Comment by ccorcos** [comment]
I've used the hoisting pattern myself:

https://github.com/ccorcos/meteor-react-instance

It works pretty well but it definitely seems like something that ought to be baked into React somehow. I prefer to use Redux now though.
